CommunityPromotionScreen/Src/NewPromotionScreenbyDefault/Classes/NPSBDP_UIArmory_PromotionHero.uc

```diff
@@ -567,35 +569,65 @@ function bool CanPurchaseAbilityEx(int Rank, int Branch, name AbilityName, out s
 		// Don't allow non hero units to purchase abilities with AP without a training center
 		// The branch is checked here so that "No Training Center" takes priority over "No
 		// class perk picked" as a reason.
-		if((UnitState.HasPurchasedPerkAtRank(Rank) || Branch >= AbilityRanks) && !UnitState.IsResistanceHero() && !CanSpendAP())
+		if ((UnitState.HasPurchasedPerkAtRank(Rank) || Branch >= AbilityRanks) && !UnitState.IsResistanceHero() && !CanSpendAP())
 		{
-			ReasonLockedText = ReasonNoTrainingCenter;
-			return false;
-		}
-
-		// Don't allow non hero units to purchase abilities on the xcom perk row before getting a rankup perk
-		if(!UnitState.HasPurchasedPerkAtRank(Rank) && !UnitState.IsResistanceHero() && Branch >= AbilityRanks )
+			strLocReasonLocked = ReasonNoTrainingCenter;
+			nReasonLocked = 'NoTrainingCenter';
+		} // Don't allow non hero units to purchase abilities on the xcom perk row before getting a rankup perk
+		else if (!UnitState.HasPurchasedPerkAtRank(Rank) && !UnitState.IsResistanceHero() && Branch >= AbilityRanks )
 		{
-			ReasonLockedText = ReasonNoClassPerkPurchased;
-			return false;
+			strLocReasonLocked = ReasonNoClassPerkPurchased;
+			nReasonLocked = 'NoClassPerkPurchased';
 		}
-	}
-
-	// Normal behaviour and hero promotion emulation (ability ranks == 0)
-	if (Rank >= UnitState.GetRank())
+	} // Normal behaviour and hero promotion emulation (ability ranks == 0)
+	else if (Rank >= UnitState.GetRank())
 	{
-		ReasonLockedText = ReasonNotHighEnoughRank;
+		strLocReasonLocked = ReasonNotHighEnoughRank;
+		nReasonLocked = 'NotHighEnoughRank';
 	}
 	else if (!CanAffordAbility(Rank, Branch))
 	{
-		ReasonLockedText = ReasonNotEnoughAP;
+		strLocReasonLocked = ReasonNotEnoughAP;
+		nReasonLocked = 'NotEnoughAP';
 	}
 	else if (!UnitState.MeetsAbilityPrerequisites(AbilityName))
 	{
-		ReasonLockedText = ReasonLacksPrerequisites;
+		strLocReasonLocked = ReasonLacksPrerequisites;
+		nReasonLocked = 'LacksPrerequisites';
 	}
 
-	return ReasonLockedText == "";
+	// Start Issue #3
+	/// Mods can listen to the 'OverrideCanPurchaseAbility' event to use their own logic 
+	/// to determine whether this particular unit should be able to unlock this particular ability.
+	/// Tuple passes:
+	/// strLocReasonLocked - localized string that will be displayed as a reason why this ability cannot be unlocked.
+	/// nReasonLocked is passed in the Tuple so mods can see *why* NPSBD's considers this ability cannot be unlocked
+	/// without having to check the strLocReasonLocked localized string.
+	/// nReasonLocked can take the following values: 
+	/// 'NoTrainingCenter', 'NoClassPerkPurchased', 'NotHighEnoughRank', 'NotEnoughAP', 'LacksPrerequisites'.
+	/// Ability will be locked if either strLocReasonLocked or nReasonLocked hold any value, so 
+	/// if a mod needs to allow this ability to be unlocked, its listener needs to "empty out" both of them.
```

Comment:
I don't understand why `nReasonLocked` is relevant as an output. If a mod wants to override the reason given, they *have* to provide `strLocReasonUnlocked`. I can understand `nReasonLocked` as an `in` argument, but it seems confusing that the mod author has to provide values for both or empty them both. And if they forget to do one of them, then `CanPurchaseAbilityEx()` automatically returns `false`.

Reply:
> I don't understand why nReasonLocked is relevant as an output.
So that a mod can lock an ability without providing a reason for it. If you don't think that behavior should be allowed, then I'm fine with making it only an `in` parameter.

Reply:
Hmmm. For user experience, I would be tempted to make it a requirement to provide a reason (nothing more frustrating to players than not knowing why they can't select something). On the other hand, I feel a bit queasy forcing something on mods that may not always be reasonable. I'm definitely conflicted on this now that you've mentioned it.

I think on balance it will be less contentious and less frustrating for modders if they can block an ability selection without providing a reason at this point. It still seems a bit confusing to have to clear these two variables, one of which is solely there (in its `out` persona) to indicate whether to block the ability selection or not.

To be honest, I prefer APIs with obvious behaviours for in and out parameters. In this case, I would simply add a boolean `out` parameter that is explicitly the sole determinant as to whether the ability selection should be blocked or not. `nReasonLocked` can then be a straightforward `in` parameter representing what the default reason is, while `strLocReasonLocked` can be set or not. Yes it's an extra "unnecessary" argument, but it improves clarity of the API, and that makes it well worth it in my view.

Reply:
Agreed, I'll adjust implementation.